### PR TITLE
Switch from using libNX_X11's deprecated XKeycodeToKeysym() function …

### PR DESCRIPTION
**Pull request by sunweaver**
…to using XGetKeyboardMapping().

 Fixes ArcticaProject/nx-libs#229.

### CODE REVIEW

**Comment by sunweaver** [comment]
@uli42: I'd be happy if you could take a look at #335 (fixing your issue #229).

File: nx-X11/programs/Xserver/hw/nxagent/Keystroke.c [hunk 1]
```diff
@@ -440,17 +447,23 @@ int nxagentCheckSpecialKeystroke(XKeyEvent *X, enum HandleEventResult *result)
...
+    free(sym);
     return 0;
   }
 
   #ifdef TEST
   fprintf(stderr, "nxagentCheckSpecialKeystroke: got code %x - state %x - sym %lx\n",
-              X -> keycode, X -> state, sym);
+              X -> keycode, X -> state, *sym);
   #endif
+  free(sym);
```
Comment (uli42) [inline nx-X11/programs/Xserver/hw/nxagent/Keystroke.c]: The manpage says "To free the storage returned by XGetKeyboardMapping, use XFree." I think using free is the way to go nowadays but mentioning this just to be sure...

File: nx-X11/programs/Xserver/hw/nxagent/Keystroke.c [hunk 2]
```diff
@@ -418,19 +424,20 @@ static enum nxagentSpecialKeystroke find_keystroke(XKeyEvent *X)
   enum nxagentSpecialKeystroke ret = KEYSTROKE_NOTHING;
 
   while (cur->stroke != KEYSTROKE_END_MARKER) {
-    if (cur->keysym == keysym && modifier_matches(cur->modifierMask, cur->modifierAltMeta, X->state)) {
+    if (cur->keysym == *keysym && modifier_matches(cur->modifierMask, cur->modifierAltMeta, X->state)) {
```
Comment (uli42) [inline nx-X11/programs/Xserver/hw/nxagent/Keystroke.c]: I _think_ as XGetKeyboardMapping returns a mapping we should use an index here to reference the data. Even if that index is always 0. Same at the other places.

Citing the mangpage:

```
The number of elements in the KeySyms list is:

       keycode_count * keysyms_per_keycode_return

       KeySym number N, counting from zero, for KeyCode K has the following index in the list, counting from zero:
       (K - first_code) * keysyms_per_code_return + N
```

**Comment by sunweaver** [comment]
@uli42: Thanks for taking a look. The point is that indexing is not trivial:

```
Keystroke.c: In function 'find_keystroke':
Keystroke.c:427:24: error: invalid type argument of unary '*' (have 'KeySym64 {aka long unsigned int}')
     if (cur->keysym == *keysym[0] && modifier_matches(cur->modifierMask, cur->modifierAltMeta, X->state)) {
                        ^~~~~~~
```

Any idea, how to cleanly index this pointer to a long unsigned int?

Mike

**Comment by sunweaver** [comment]
Oh, actually this works:

```
diff --git a/nx-X11/programs/Xserver/hw/nxagent/Keystroke.c b/nx-X11/programs/Xserver/hw/nxagent/Keystroke.c
index 67c271ea..dd2780fc 100644
--- a/nx-X11/programs/Xserver/hw/nxagent/Keystroke.c
+++ b/nx-X11/programs/Xserver/hw/nxagent/Keystroke.c
@@ -424,7 +424,7 @@ static enum nxagentSpecialKeystroke find_keystroke(XKeyEvent *X)
   enum nxagentSpecialKeystroke ret = KEYSTROKE_NOTHING;
 
   while (cur->stroke != KEYSTROKE_END_MARKER) {
-    if (cur->keysym == *keysym && modifier_matches(cur->modifierMask, cur->modifierAltMeta, X->state)) {
+    if (cur->keysym == keysym[0] && modifier_matches(cur->modifierMask, cur->modifierAltMeta, X->state)) {
 
       free(keysym);
       return cur->stroke;
```

**Comment by sunweaver** [comment]
@uli42: ok· I force-pushed an update. The Xfree() function has been deprecated all over nx-libs now.

The index thingy has been implemented in the most recent push. Please check once more and approve the PR, if ok.

**Review by uli42 — COMMENTED** [review]
this is ok.

In theory we can remove XKeycodeToKeysym() then, but libX11 upstream still delivers it.

**Comment by Ionic** [comment]
Uh... but `*keysim` and `keysim[0]` are completely equivalent, with `*keysim` being a bit more idiomatic, I guess.

**Comment by Ionic** [comment]
For added fun, `0[keysim]` is equally legit and does the same thing, as what really happens is just pointer arithmetics (pointer value + 0 or 0 + pointer value in this case) and a dereference.

**Comment by uli42** [comment]
On Sat, Feb 18, 2017 at 10:32 AM, Mihai Moldovan
<notifications@github.com> wrote:
>
> For added fun, 0[keysim] is equally legit and does the same thing, as what really happens is just pointer arithmetics (pointer value + 0 or 0 + pointer value in this case) and a dereference.

Of course this is the same. But using an index makes it more readable
IMHO. And it show that the function returns an array.

**Pull request merged**